lib/puppet/indirector/catalog/rest.rb

```diff
@@ -3,4 +3,51 @@
 
 class Puppet::Resource::Catalog::Rest < Puppet::Indirector::REST
   desc "Find resource catalogs over HTTP via REST."
+
+  def find(request)
+    return super unless use_http_client?
+
+    # URL encoded facts and facts_format are passed as indirector
+    # request options, so we have to reverse that (unescape, then parse),
+    # and pass a facts object to the http client.
+    format = request.options[:facts_format]
+    if format
+      formatter = Puppet::Network::FormatHandler.format_for(format)
+      facts = formatter.intern(Puppet::Node::Facts, CGI.unescape(request.options[:facts]))
+    else
+      facts = Puppet::Node::Facts.new(request.key, environment: request.environment.to_s)
+    end
+
+    checksum_type = if request.options[:checksum_type]
+                      request.options[:checksum_type].split('.')
+                    else
+                      Puppet[:supported_checksum_types]
+                    end
+
+    session = Puppet.lookup(:http_session)
+    api = session.route_to(:puppet)
+    catalog = api.get_catalog(
```

Comment:
I feel like we should have named this method `post_catalog` rather than `get_catalog`. It's a bit confusing.

Also, why are we using post to retrieve the catalog? That seems like it would be better as a get request? I think I missed this when we initial got this method in

Reply:
Agree `post_catalog` is better. Previously, puppet defaults to issuing a GET request, but if the HTTP request line is too long, then it switches to a POST. Since facts are submitted as query parameters, we never issued a GET in practice. I'll update the service method and this call.